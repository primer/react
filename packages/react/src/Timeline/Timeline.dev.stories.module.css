.TimelineWithMarginLeft {
  margin-left: var(--base-size-16);
}

.TimelineItemWithPadding {
  padding-top: var(--base-size-8);
  padding-bottom: var(--base-size-4);
}

.TimelineBadgeCanvas {
  background-color: var(--bgColor-default);
}

.TimelineBadgeDanger {
  background-color: var(--bgColor-danger-emphasis);
}

.TimelineBadgeAccent {
  margin-top: var(--base-size-8);
  background-color: var(--bgColor-accent-emphasis);
}

.TimelineBodySubtle {
  color: var(--fgColor-muted);
  font-size: var(--text-body-size-medium);
  width: 100%;
  padding-right: var(--base-size-20);
  display: flex;
  flex-direction: column;
}

.TimelineBodyDefault {
  color: var(--fgColor-default);
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.TimelineBreakThick {
  border-width: var(--borderWidth-thick);
}