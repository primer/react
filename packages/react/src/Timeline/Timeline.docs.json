{
  "id": "timeline",
  "name": "Timeline",
  "status": "alpha",
  "a11yReviewed": false,
  "stories": [],
  "importPath": "@primer/react",
  "props": [
    {
      "name": "clipSidebar",
      "type": "boolean",
      "description": "Hides the sidebar above the first Timeline.Item and below the last Timeline.Item."
    },
    {
      "name": "sx",
      "type": "SystemStyleObject"
    }
  ],
  "subcomponents": [
    {
      "name": "Timeline.Item",
      "props": [
        {
          "name": "condensed",
          "type": "boolean",
          "description": "Reduces vertical padding and removes background from an item's badge."
        },
        {
          "name": "sx",
          "type": "SystemStyleObject"
        }
      ]
    },
    {
      "name": "Timeline.Badge",
      "props": [
        {
          "name": "sx",
          "type": "SystemStyleObject"
        }
      ]
    },
    {
      "name": "Timeline.Body",
      "props": [
        {
          "name": "sx",
          "type": "SystemStyleObject"
        }
      ]
    },
    {
      "name": "Timeline.Break",
      "props": [
        {
          "name": "sx",
          "type": "SystemStyleObject"
        }
      ]
    }
  ]
}
