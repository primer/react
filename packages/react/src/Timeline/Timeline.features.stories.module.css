.BadgeWithDoneBackground {
  background-color: var(--bgColor-done-emphasis);
}

.LinkWithBoldStyle {
  font-weight: var(--base-text-weight-semibold);
  color: var(--fgColor-default);
  margin-right: var(--base-size-4);
}

.LinkWithBoldStyle:hover {
  color: var(--fgColor-accent);
}

.GitMergeIcon {
  color: var(--fgColor-onEmphasis);
}
