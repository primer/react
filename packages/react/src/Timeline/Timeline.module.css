.Timeline {
  display: flex;
  flex-direction: column;

  &:where([data-clip-sidebar]) {
    .TimelineItem:first-child {
      padding-top: 0;
    }

    .TimelineItem:last-child {
      padding-bottom: 0;
    }
  }
}

.TimelineItem {
  position: relative;
  display: flex;
  padding: var(--base-size-16) 0;
  margin-left: var(--base-size-16);

  &::before {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    display: block;
    width: 2px;
    content: '';
    /* stylelint-disable-next-line primer/colors */
    background-color: var(--borderColor-muted);
  }

  &:where([data-condensed]) {
    padding-top: var(--base-size-4);
    padding-bottom: 0;

    &:last-child {
      padding-bottom: var(--base-size-16);
    }

    .TimelineBadge {
      height: 16px;
      margin-top: var(--base-size-8);
      margin-bottom: var(--base-size-8);
      color: var(--fgColor-muted);
      background-color: var(--bgColor-default);
      border: 0;
    }
  }
}

.TimelineBadgeWrapper {
  position: relative;
  z-index: 1;
}

.TimelineBadge {
  display: flex;
  width: 32px;
  height: 32px;
  margin-right: var(--base-size-8);
  /* stylelint-disable-next-line primer/spacing */
  margin-left: -15px;
  flex-shrink: 0;
  overflow: hidden;
  color: var(--fgColor-muted);

  /* TODOl not quite sure if this is the correct migration for this line */
  background-color: var(--timelineBadge-bgColor);
  /* stylelint-disable-next-line primer/colors */
  border-color: var(--bgColor-default);
  border-style: solid;
  border-width: var(--borderWidth-thick);
  border-radius: 50%;
  align-items: center;
  justify-content: center;
}

.TimelineBody {
  min-width: 0;
  max-width: 100%;
  margin-top: var(--base-size-4);
  font-size: var(--text-body-size-medium);
  color: var(--fgColor-muted);
  flex: auto;
}

.TimelineBreak {
  position: relative;
  z-index: 1;
  height: var(--base-size-24);
  margin: 0;
  margin-bottom: calc(-1 * var(--base-size-16));
  margin-left: 0;
  background-color: var(--bgColor-default);
  border: 0;
  border-top: var(--borderWidth-thicker) solid var(--borderColor-default);
}
