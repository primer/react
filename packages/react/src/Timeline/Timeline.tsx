import clsx from 'clsx'
import React from 'react'
import styled, {css} from 'styled-components'
import Box from '../Box'
import {get} from '../constants'
import type {SxProp} from '../sx'
import sx from '../sx'
import type {ComponentProps} from '../utils/types'

const Timeline = styled.div<{clipSidebar?: boolean} & SxProp>`
  display: flex;
  flex-direction: column;
  ${props =>
    props.clipSidebar &&
    css`
      .Timeline-Item:first-child {
        padding-top: 0;
      }

      .Timeline-Item:last-child {
        padding-bottom: 0;
      }
    `}

  ${sx};
`

type StyledTimelineItemProps = {condensed?: boolean} & SxProp

const TimelineItem = styled.div.attrs<StyledTimelineItemProps>(props => ({
  className: clsx('Timeline-Item', props.className),
}))<StyledTimelineItemProps>`
  display: flex;
  position: relative;
  padding: ${get('space.3')} 0;
  margin-left: ${get('space.3')};

  &::before {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    display: block;
    width: 2px;
    content: '';
    background-color: ${get('colors.border.muted')};
  }

  ${props =>
    props.condensed &&
    css`
      padding-top: ${get('space.1')};
      padding-bottom: 0;
      &:last-child {
        padding-bottom: ${get('space.3')};
      }

      .TimelineItem-Badge {
        height: 16px;
        margin-top: ${get('space.2')};
        margin-bottom: ${get('space.2')};
        color: ${get('colors.fg.muted')};
        background-color: ${get('colors.canvas.default')};
        border: 0;
      }
    `}

  ${sx};
`

export type TimelineBadgeProps = {children?: React.ReactNode} & SxProp

const TimelineBadge = (props: TimelineBadgeProps) => {
  return (
    <Box position="relative" zIndex={1}>
      <Box
        display="flex"
        className="TimelineItem-Badge"
        flexShrink={0}
        borderRadius="50%"
        borderWidth="2px"
        borderStyle="solid"
        borderColor="canvas.default"
        overflow="hidden"
        color="fg.muted"
        bg="timeline.badgeBg"
        width="32px"
        height="32px"
        mr={2}
        ml="-15px"
        alignItems="center"
        justifyContent="center"
        sx={props.sx}
      >
        {props.children}
      </Box>
    </Box>
  )
}

const TimelineBody = styled.div<SxProp>`
  min-width: 0;
  max-width: 100%;
  margin-top: ${get('space.1')};
  color: ${get('colors.fg.muted')};
  flex: auto;
  font-size: ${get('fontSizes.1')};
  ${sx};
`

const TimelineBreak = styled.div<SxProp>`
  position: relative;
  z-index: 1;
  height: 24px;
  margin: 0;
  margin-bottom: -${get('space.3')};
  margin-left: 0;
  background-color: ${get('colors.canvas.default')};
  border: 0;
  border-top: ${get('space.1')} solid ${get('colors.border.default')};
  ${sx};
`

TimelineItem.displayName = 'Timeline.Item'

TimelineBadge.displayName = 'Timeline.Badge'

TimelineBody.displayName = 'Timeline.Body'

TimelineBreak.displayName = 'Timeline.Break'

export type TimelineProps = ComponentProps<typeof Timeline>
export type TimelineItemsProps = ComponentProps<typeof TimelineItem>
export type TimelineBodyProps = ComponentProps<typeof TimelineBody>
export type TimelineBreakProps = ComponentProps<typeof TimelineBreak>
export default Object.assign(Timeline, {
  Item: TimelineItem,
  Badge: TimelineBadge,
  Body: TimelineBody,
  Break: TimelineBreak,
})
