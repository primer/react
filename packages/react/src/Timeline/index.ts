export {default} from './Timeline'
export type {
  TimelineProps,
  TimelineItemsProps,
  TimelineItemProps,
  TimelineBadgeProps,
  TimelineBodyProps,
  TimelineBreakProps,
} from './Timeline'
