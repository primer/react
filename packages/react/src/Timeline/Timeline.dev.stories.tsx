import React from 'react'
import type {Meta} from '@storybook/react'
import type {ComponentProps} from '../utils/types'
import Timeline from './Timeline'
import Octicon from '../Octicon'
import {GitCommitIcon} from '@primer/octicons-react'

export default {
  title: 'Components/Timeline/Dev',
  component: Timeline,
  subcomponents: {
    'Timeline.Item': Timeline.Item,
    'Timeline.Badge': Timeline.Badge,
    'Timeline.Body': Timeline.Body,
    'Timeline.Break': Timeline.Break,
  },
} as Meta<ComponentProps<typeof Timeline>>

export const SxProps = () => (
  <Timeline
    sx={{
      ml: 4,
    }}
    style={{border: '1px solid red'}}
  >
    <Timeline.Item
      sx={{
        pt: 2,
        pb: 1,
      }}
      style={{border: '1px solid pink'}}
    >
      <Timeline.Badge
        sx={{
          backgroundColor: 'canvas.default',
        }}
      >
        <Octicon icon={GitCommitIcon} aria-label="Commit" />
      </Timeline.Badge>
      <Timeline.Body
        sx={{
          color: 'fg.subtle',
          fontSize: '14px',
          width: '100%',
          paddingRight: 5,
          display: 'flex',
          flexDirection: 'column',
        }}
        className="mt-0"
        style={{border: '1px solid green'}}
      >
        This is a message
      </Timeline.Body>
    </Timeline.Item>
    <Timeline.Item color="gray">
      <Timeline.Badge
        sx={{
          bg: 'danger.emphasis',
        }}
      >
        <Octicon icon={GitCommitIcon} aria-label="Commit" />
      </Timeline.Badge>
      <Timeline.Body
        sx={{
          color: 'fg.default',
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'space-between',
        }}
      >
        This is a message
      </Timeline.Body>
    </Timeline.Item>
    <Timeline.Break
      sx={{
        borderWidth: '2px',
      }}
      style={{border: '1px solid gray'}}
    />
    <Timeline.Item condensed={true}>
      <Timeline.Badge
        sx={{
          mt: 2,
          bg: 'accent.emphasis',
        }}
      >
        <Octicon icon={GitCommitIcon} aria-label="Commit" />
      </Timeline.Badge>
      <Timeline.Body>This is a message</Timeline.Body>
    </Timeline.Item>
  </Timeline>
)
