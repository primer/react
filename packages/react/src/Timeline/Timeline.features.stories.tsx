import React from 'react'
import type {Meta} from '@storybook/react'
import type {ComponentProps} from '../utils/types'
import Timeline from './Timeline'
import Octicon from '../Octicon'
import {GitBranchIcon, GitCommitIcon, GitMergeIcon} from '@primer/octicons-react'
import Link from '../Link'

export default {
  title: 'Components/Timeline/Features',
  component: Timeline,
  subcomponents: {
    'Timeline.Item': Timeline.Item,
    'Timeline.Badge': Timeline.Badge,
    'Timeline.Body': Timeline.Body,
    'Timeline.Break': Timeline.Break,
  },
} as Meta<ComponentProps<typeof Timeline>>

export const ClipSidebar = () => (
  <Timeline clipSidebar>
    <Timeline.Item>
      <Timeline.Badge>
        <Octicon icon={GitCommitIcon} />
      </Timeline.Badge>
      <Timeline.Body>This is a message</Timeline.Body>
    </Timeline.Item>
  </Timeline>
)

export const CondensedItems = () => (
  <Timeline>
    <Timeline.Item condensed>
      <Timeline.Badge>
        <Octicon icon={GitCommitIcon} />
      </Timeline.Badge>
      <Timeline.Body>This is a message</Timeline.Body>
    </Timeline.Item>
    <Timeline.Item condensed>
      <Timeline.Badge>
        <Octicon icon={GitCommitIcon} />
      </Timeline.Badge>
      <Timeline.Body>This is a message</Timeline.Body>
    </Timeline.Item>
  </Timeline>
)

export const TimelineBreak = () => (
  <Timeline>
    <Timeline.Item>
      <Timeline.Badge sx={{bg: 'done.emphasis'}}>
        <Octicon icon={GitMergeIcon} color="fg.onEmphasis" />
      </Timeline.Badge>
      <Timeline.Body>This is a message</Timeline.Body>
    </Timeline.Item>
    <Timeline.Break />
    <Timeline.Item>
      <Timeline.Badge>
        <Octicon icon={GitBranchIcon} />
      </Timeline.Badge>
      <Timeline.Body>This is a message</Timeline.Body>
    </Timeline.Item>
  </Timeline>
)

export const WithInlineLinks = () => (
  <Timeline>
    <Timeline.Item>
      <Timeline.Badge>
        <Octicon icon={GitCommitIcon} />
      </Timeline.Badge>
      <Timeline.Body>
        <Link href="#" sx={{fontWeight: 'bold', color: 'fg.default', mr: 1}} muted>
          Monalisa
        </Link>
        enabled auto-merge (squash)
      </Timeline.Body>
    </Timeline.Item>
  </Timeline>
)
