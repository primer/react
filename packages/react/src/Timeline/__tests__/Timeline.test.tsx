import {render as HTMLRender} from '@testing-library/react'
import axe from 'axe-core'
import {render, rendersClass, behavesAsComponent, checkExports} from '../../utils/testing'

import React from 'react'
import Timeline from '..'

describe('Timeline', () => {
  behavesAsComponent({Component: Timeline, options: {skipAs: true}})

  checkExports('Timeline', {
    default: Timeline,
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Timeline />)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('renders with clipSidebar prop', () => {
    expect(render(<Timeline clipSidebar />)).toMatchSnapshot()
  })

  it('should support `className` on the outermost element', () => {
    expect(HTMLRender(<Timeline className={'test-class-name'} />).container.firstChild).toHaveClass('test-class-name')
  })
})

describe('Timeline.Item', () => {
  behavesAsComponent({Component: Timeline.Item, options: {skipAs: true}})

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Timeline.Item />)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('renders with condensed prop', () => {
    expect(render(<Timeline.Item condensed />)).toMatchSnapshot()
  })

  it('adds the Timeline-Item class', () => {
    expect(rendersClass(<Timeline.Item />, 'Timeline-Item')).toEqual(true)
  })

  it('should support `className` on the outermost element', () => {
    expect(HTMLRender(<Timeline.Item className={'test-class-name'} />).container.firstChild).toHaveClass(
      'test-class-name',
    )
  })
})

describe('Timeline.Badge', () => {
  behavesAsComponent({Component: Timeline.Badge, options: {skipAs: true}})

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Timeline.Badge />)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('should support `className` on the outermost element', () => {
    expect(HTMLRender(<Timeline.Badge className={'test-class-name'} />).container.firstChild?.firstChild).toHaveClass(
      'test-class-name',
    )
  })
})

describe('Timeline.Body', () => {
  behavesAsComponent({Component: Timeline.Badge, options: {skipAs: true}})

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Timeline.Body />)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('should support `className` on the outermost element', () => {
    expect(HTMLRender(<Timeline.Body className={'test-class-name'} />).container.firstChild).toHaveClass(
      'test-class-name',
    )
  })
})

describe('Timeline.Break', () => {
  behavesAsComponent({Component: Timeline.Badge, options: {skipAs: true}})

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Timeline.Break />)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('should support `className` on the outermost element', () => {
    expect(HTMLRender(<Timeline.Break className={'test-class-name'} />).container.firstChild).toHaveClass(
      'test-class-name',
    )
  })
})
