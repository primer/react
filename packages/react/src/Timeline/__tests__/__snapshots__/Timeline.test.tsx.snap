// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline renders with clipSidebar prop 1`] = `
<div
  className="Timeline"
  data-clip-sidebar=""
/>
`;

exports[`Timeline.Item renders with condensed prop 1`] = `
<div
  className="Timeline-Item TimelineItem"
  data-condensed=""
/>
`;
